File: backend/app/api/v1/llm.py
"""LLM utility endpoints — exposes the host Ollama's available models so the
frontend doesn't have to hardcode a list."""
import logging

from fastapi import APIRouter, HTTPException

from app.core.deps import CurrentUser
from app.schemas.review import OllamaModel, OllamaModelsResponse
from app.services import ollama_service

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/llm", tags=["llm"])


@router.get(
    "/models",
    response_model=OllamaModelsResponse,
    summary="List models available on the host Ollama (proxies /api/tags)",
)
async def list_models(_user: CurrentUser):
    try:
        raw = await ollama_service.list_models()
    except ollama_service.OllamaError as e:
        logger.warning("Could not reach Ollama: %s", e)
        raise HTTPException(
            status_code=502,
            detail=f"Could not reach the local LLM service: {e}",
        )
    out: list[OllamaModel] = []
    for m in raw:
        details = m.get("details") or {}
        out.append(
            OllamaModel(
                name=m.get("name") or m.get("model") or "",
                parameter_size=details.get("parameter_size"),
                family=details.get("family"),
                size_bytes=m.get("size"),
            )
        )
    return OllamaModelsResponse(models=out)

File: frontend/src/api/llm.ts
import { api } from "./client";

export interface OllamaModel {
  name: string;
  parameter_size: string | null;
  family: string | null;
  size_bytes: number | null;
}

export interface OllamaModelsResponse {
  models: OllamaModel[];
}

export async function listModels(): Promise<OllamaModelsResponse> {
  const res = await api.get<OllamaModelsResponse>("/llm/models");
  return res.data;
}

/**
 * Format a model for display in dropdowns.
 *   gemma4:latest (8.0B, gemma4) -> "gemma4:latest — 8.0B"
 */
export function formatModelLabel(m: OllamaModel): string {
  const parts: string[] = [m.name];
  if (m.parameter_size) parts.push(m.parameter_size);
  return parts.join(" — ");
}
